### DISCUSSION TITLE
vs1053

### DISCUSSION THREAD

**Discussion by andynvkz** [pschatzmann/ESP32-A2DP#448] [Q&A] [answered]
You indicated that the stream is audi PCM, but vs1053 does not perceive it

**Comment by pschatzmann** [comment, marked as answer]
This has nothing to do with A2DP and as far as I know the vs1053 does not support PCM.

I converted this into a discussion because there is clearly no bug in my library.
Did you try to encode the output into a WAV format ?

**Reply by andynvkz** [reply]
I did not encode the output to wav format, I thought that the data is exactly what comes out, vs1053 (https://cdn-shop.adafruit.com/datasheets/vs1053.pdf)

**Comment by pschatzmann** [comment]
Please read the specification again: It indicates that PCM and ADPCM is supported by providing WAV!

After all that's exactly what my other projects are about e.g.
- https://github.com/pschatzmann/arduino-audio-tools
- https://github.com/pschatzmann/arduino-vs1053

**Comment by andynvkz** [comment]
Tell me how and how can I convert this format so that it would be accepted by vs1053?

**Comment by andynvkz** [comment]
It would be ideal if they added support for this chip to the library

**Comment by pschatzmann** [comment]
That does not make any sense at all. The A2DP library is part of the AudioTools which provides exactly what you are asking for

**Comment by andynvkz** [comment]
I just don't understand how to use it

**Comment by andynvkz** [comment]
how do i get the vs1053 stream to play? from read_data_stream

**Comment by pschatzmann** [comment]
Use an EncodedAudioStream and define the output to be a VS1053Stream, also provide a WAV Encoder so that the output is converted to WAV.

In the data callback function you just write to the EncodedAudioStream.

Don't forget to open the streams properly...

If you double check the VS1053Stream API, there is even a short cut where the above is done  automatically.

**Reply by andynvkz** [reply]
I try your example: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/sandbox/basic-a2dp-vs1053/basic-a2dp-vs1053.ino , can't get sound, vs1053 connected to VSPI

**Reply by andynvkz** [reply]
Started
[I] AudioStreams.h : 893 - bool audio_tools::QueueStream<T>::begin() [with T = unsigned char]
[I] VS1053Stream.h : 106 - bool audio_tools::VS1053Stream::begin(audio_tools::VS1053Config)
[I] AudioTypes.h : 88 - sample_rate: 44100
[I] AudioTypes.h : 89 - channels: 2
[I] AudioTypes.h : 90 - bits_per_sample: 16
[I] VS1053Stream.h : 115 - is_encoded_data: false
[I] VS1053Stream.h : 116 - is_midi: false
[I] VS1053Stream.h : 117 - cs_pin: 32
[I] VS1053Stream.h : 118 - dcs_pin: 33
[I] VS1053Stream.h : 119 - dreq_pin: 35
[I] VS1053Stream.h : 120 - reset_pin: -1
[I] VS1053Stream.h : 121 - cs_sd_pin: -1
[I] VS1053Stream.h : 128 - SPI.begin()
[I] VS1053Stream.h : 337 - bool audio_tools::VS1053Stream::beginTx()
[I] CodecWAV.h : 482 - sample_rate: 44100
[I] CodecWAV.h : 483 - channels: 2
[I] CodecWAV.h : 492 - is_streamed! because length is 2147418112
E (26163) gpio: gpio_set_level(226): GPIO output gpio_num error
[I] CodecWAV.h : 531 - Writing Header
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
VS1053 - VS1053 not properly installed!
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops

**Reply by pschatzmann** [reply]
I think you just ignored the readme of that directory!
I guess the Error is coming from the undefined reset pin, though I would not understand where this could come from.
If you see these copy lines, that's a good sign and inicates that the vs1053 is getting the data.

I am not sure why I used the buffer, so you might want to try a version without and one which uses a synchronized buffer implementation to avoid any concurrency conflicts.

I also think there is a bug in the read_data_stream method. Compare it with one of the tested examples!

I updated the sandbox example, but I let you pick up from here because I don't have any hardware to test

**Reply by andynvkz** [reply]
I tried to insert a wav title at the beginning of the buffer and the sound appeared, but it was so quiet...

**Reply by pschatzmann** [reply]
So, just bump up the volume on your phone!

**Reply by andynvkz** [reply]
it's not the volume, something is wrong with the PCM format

**Reply by pschatzmann** [reply]
So, did you check if the write was processed correctly ? 
I would have expected that you extend the error handling in this case...

**Reply by andynvkz** [reply]
in the title, I indicated that the data is 44100 Hz 16-bit stereo, right? and format PCM or ADPCM?

**Reply by pschatzmann** [reply]
The data format is 16 bit PCM data with 2 channels usually at 44100 hz: You can find this information in the README, that you were supposed to read before starting a discussion!
I suggest that you also add a check after starting a2dp to verify the effective sample rate with the API

I don't quite understand what you mean with title....

**Reply by andynvkz** [reply]
RIFF header

**Comment by andynvkz** [comment]
Thank you, I will try

**Comment by andynvkz** [comment]
Everything worked out great for me, YouTube transmits sound, but how can I make sure that the sound is transmitted when making a call?
